opm/parser/eclipse/Utility/Functional.hpp

```diff
@@ -0,0 +1,214 @@
+     * functionality and higher order functions inspiried by functional
+     * languages (in particular Haskell) and modern C++. The goal is to provide
+     * lightweight features that reduce boilerplate and make code more
+     * declarative.
+     */
+
+    /*
+     * map :: (a -> b) -> [a] -> [b]
+     *
+     * maps the elements [a] of the passed container C to [b], by using the
+     * passed function f :: a -> b. Works like map in haskell, lisp, python etc.
+     *
+     * C can be any foreach-compatible container (that supports .begin,
+     * .end), but will always return a vector.
+     *
+     * F can be any Callable, that is both function pointer,
+     * operator()-providing class or std::function, including lambdas. F is
+     * typically passed by reference. F must be unary of type A (which must
+     * match what C::const_iterator::operator* returns) and have return
+     * type B (by value).
+     *
+     * In short, this function deal with vector allocation, resizing and
+     * population based on some function f.
+     *
+     * fun::map( f, vec ) is equivalent to:
+     *  vector dst;
+     *  for( auto& x : vec ) dst.push_back( f( x ) );
+     *  return dst;
+     *
+     * The behaviour is undefined if F has any side effects.
+     *
+     * --
+     *
+     * int plus1( int x ) { return x + 1; }
+     * base_vec = { 0, 1, 2, 3, 4 };
+     * vec = fun::map( &plus1, base_vec );
+     *
+     * vec => { 1, 2, 3, 4, 5 }
+     *
+     * --
+     *
+     * int mul2 = []( int x ) { return x * 2; };
+     * base_vec = { 0, 1, 2, 3, 4 };
+     * vec = fun::map( mul2, base_vec );
+     *
+     * vec => { 0, 2, 4, 6, 8 };
+     *
+     */
+    template< typename F, typename C >
+    std::vector< typename std::result_of< F( typename C::const_iterator::value_type& ) >::type >
+    map( F f, const C& src ) {
+        using A = typename C::const_iterator::value_type;
+        using B = typename std::result_of< F( A& ) >::type;
+        std::vector< B > ret;
+        ret.reserve( src.size() );
+
+        std::transform( src.begin(), src.end(), std::back_inserter( ret ), f );
+        return ret;
+    }
+
+    /*
+     * concat :: [[a]] -> [a]
+     *
+     * A primitive concat taking a vector of vectors, flattened into a
+     * single 1 dimensional vector. Moves all the elements so no unecessary
+     * copies are done.
+     *
+     * vec = { { 1 }, { 2, 2 }, { 3, 3, 3 } }
+     * cvec = concat( vec ) => { 1, 2, 2, 3, 3, 3 }
+     */
+    template< typename A >
+    std::vector< A > concat( std::vector< std::vector< A > >&& src ) {
+        const auto size = std::accumulate( src.begin(), src.end(), 0,
+                []( std::size_t acc, const std::vector< A >& x ) {
+                    return acc + x.size();
+                    }
+                );
+
+        std::vector< A > dst;
+        dst.reserve( size );
+
+        for( auto& x : src )
+            std::move( x.begin(), x.end(), std::back_inserter( dst ) );
```

Comment:
Nine times out of ten it's the std::move from utility you're looking for, as std::move(begin, end, output) is rather specific to the algorithms library. Confusing maybe, but makes sense in my opinion. It's a *very* useful feature however, as demonstrated here.